File: src/Callbacks.hs
module Callbacks where

import Data.IORef
import qualified Data.Set as S
import Graphics.UI.GLFW
import Graphics.Rendering.OpenGL
import System.Exit

data Camera = Camera (Vector3 GLdouble) (GLdouble, GLdouble)
data Model  = Model (Int, Int) (S.Set Key)

myKeyCallback :: IORef Model -> KeyCallback
myKeyCallback model (SpecialKey ESC) Press =
    closeWindow >> terminate >> exitSuccess
myKeyCallback model key state = do
    Model regionIx pressedKeys <- get model
    if state == Press
       then model $= Model regionIx (S.insert key pressedKeys)
       else model $= Model regionIx (S.delete key pressedKeys)

myMouseCallback :: IORef Camera -> MousePosCallback
myMouseCallback camera (Position x y) = do
    Camera pos (rotX, rotY) <- get camera
    Size w h <- get windowSize

    let midX = w `quot` 2
        midY = h `quot` 2
        dx = fromIntegral $ x - midX
        dy = fromIntegral $ y - midY

    camera   $= Camera pos (clamp (-90) 90 (rotX + dy), (roll (-180) 180 (rotY + dx)))
    mousePos $= Position midX midY

rad2deg :: Floating a => a -> a
{-# INLINE rad2deg #-}
rad2deg rad = rad * 180 / pi

deg2rad :: Floating a => a -> a
{-# INLINE deg2rad #-}
deg2rad deg = deg * pi / 180

clamp :: Ord a => a -> a -> a -> a
{-# INLINE clamp #-}
clamp lo hi x = min (max x lo) hi

roll :: Ord a => a -> a -> a -> a
{-# INLINE roll #-}
roll lo hi x
    | x < lo = hi
    | x > hi = lo
    | otherwise = x

File: c/callbacks.cpp
#include "callbacks.hpp"

extern Camera g_camera;

void GLFWCALL keyCallback(const int key, const int action)
{
    if (action == GLFW_PRESS) {
        switch (key) {
            case 'W':
                g_camera.movement |= MOVE_FORWARD; break;
            case 'S':
                g_camera.movement |= MOVE_BACKWARD; break;
            case 'A':
                g_camera.movement |= MOVE_LEFT; break;
            case 'D':
                g_camera.movement |= MOVE_RIGHT; break;
            case '-':
                g_camera.movement |= MOVE_DOWN; break;
            case '=':
                g_camera.movement |= MOVE_UP; break;
            default:
                break;
        }
    } else {
        switch (key) {
            case 'W':
                g_camera.movement &= ~MOVE_FORWARD; break;
            case 'S':
                g_camera.movement &= ~MOVE_BACKWARD; break;
            case 'A':
                g_camera.movement &= ~MOVE_LEFT; break;
            case 'D':
                g_camera.movement &= ~MOVE_RIGHT; break;
            case '-':
                g_camera.movement &= ~MOVE_DOWN; break;
            case '=':
                g_camera.movement &= ~MOVE_UP; break;
            default:
                break;
        }
    }
}

template <typename T>
static inline void clamp(T& x, const T& low, const T& high)
{
    x = x < low ? low : (x > high ? high : x);
}

template <typename T>
static inline void roll(T& x, const T& low, const T& high)
{
    x = x < low ? high : (x > high ? low : x);
}

void GLFWCALL mousePosCallback(const int x, const int y)
{
    int w, h;
    glfwGetWindowSize(&w, &h);

    int midX = w * 0.5f;
    int midY = h * 0.5f;

    g_camera.rotX += y - midY;
    g_camera.rotY += x - midX;

    clamp(g_camera.rotX, -90.0f, 90.0f);
    roll(g_camera.rotY, -180.0f, 180.0f);

    glfwSetMousePos(midX, midY);
}

void GLFWCALL windowSizeCallback(int w, int h)
{
    glViewport(0, 0, w, h);
}
